File: events/2015/17/solution.py
import sys
import json
import time
from datetime import timedelta, datetime
from pathlib import Path

# Add workspace root to path for imports
workspace_root = Path(__file__).parent.parent.parent.parent
sys.path.insert(0, str(workspace_root))

from utils.python.types import AOCDayResults, AOCTimestamp, AOCPartResult
from utils.python.input import readInputAsArray
from utils.python.performance import measurePerformance

def count_permutations(container_sizes: list[int], target: int) -> int:
    # Create array to track the number of ways to make any value between zero and the target
    ways_to_make_sum = [0] * (target + 1)
    # There is only one way to make a value of 0, use no containers!
    ways_to_make_sum[0] = 1
    
    # Loop over each container
    for container in container_sizes:
        # Traverse backwards to avoid using same container twice
        for i in range(target, container - 1, -1):
            ways_to_make_sum[i] += ways_to_make_sum[i - container]
    
    return ways_to_make_sum[target]


def count_min_permutations(container_sizes: list[int], target: int) -> dict[str, int]:
    # First pass: find minimum containers needed
    min_containers_needed = [float('inf')] * (target + 1)
    min_containers_needed[0] = 0
    
    for container in container_sizes:
        for i in range(target, container - 1, -1):
            min_containers_needed[i] = min(
                min_containers_needed[i],
                min_containers_needed[i - container] + 1
            )
    
    min_containers = min_containers_needed[target]
    
    # Second pass: count permutations using exactly min_containers
    ways_to_make_sum_with_n_containers = [
        [0] * (min_containers + 1) for _ in range(target + 1)
    ]
    ways_to_make_sum_with_n_containers[0][0] = 1
    
    for container in container_sizes:
        for sum_val in range(target, container - 1, -1):
            for num_containers in range(min_containers, 0, -1):
                ways_to_make_sum_with_n_containers[sum_val][num_containers] += \
                    ways_to_make_sum_with_n_containers[sum_val - container][num_containers - 1]
    
    return {
        'min_containers': min_containers,
        'permutation_count': ways_to_make_sum_with_n_containers[target][min_containers]
    }

# ----------------------------------------------------------------------------------------------------
# | Part 1
# ----------------------------------------------------------------------------------------------------
def part1(input: list[str]) -> int:
  container_sizes: list[int] = []
  total = 150

  for item in input:
    if len(item) == 0:
      continue
    container_sizes.append(int(item))

  permutations = count_permutations(container_sizes, total)
  return permutations

# ----------------------------------------------------------------------------------------------------
# | Part 2
# ----------------------------------------------------------------------------------------------------
def part2(input: list[str]) -> int:
  container_sizes: list[int] = []
  total = 150

  for item in input:
    if len(item) == 0:
      continue
    container_sizes.append(int(item))

  permutations = count_min_permutations(container_sizes, total)
  return permutations["permutation_count"]

# ----------------------------------------------------------------------------------------------------
# | Main Function
# ----------------------------------------------------------------------------------------------------
def main():
  # Get the arguments
  args = sys.argv[1:]
  fileName = args[0] if args else "input.txt"

  # Start the timer
  start_timestamp = datetime.now().isoformat()
  start_time = time.monotonic()

  # Read the input to an array of strings
  input = readInputAsArray(fileName)

  ## Run the parts
  p1Result = measurePerformance(part1, input)
  p2Result = measurePerformance(part2, input)

  # End the timer
  end_time = time.monotonic()
  end_timestamp = datetime.now().isoformat()
  duration = timedelta(seconds=end_time - start_time).total_seconds()*1000000000 # Convert to nanoseconds

  # Return the results
  return AOCDayResults(
    Year=2015,
    Day=17,
    Part1=p1Result,
    Part2=p2Result,
    Duration=duration,
    Timestamp=AOCTimestamp(
      Start=start_timestamp,
      End=end_timestamp,
    ),
  )



if __name__ == "__main__":
  results = main()
  print(json.dumps(results))

File: events/2015/17/solution.ts
import { AOCDayResults } from "@repo/utils-ts/returnType";
import { readInputAsArray } from "@repo/utils-ts/readInput";
import { measurePerformance } from "@repo/utils-ts/measurePerformance";

function countPermutations(containerSizes: number[], target: number): number {
  // Create array to track the number of ways to make any value between zero and the target
  const waysToMakeSum: number[] = new Array(target + 1).fill(0);

  // There is onle one way to make a value of 0, use no containers!
  waysToMakeSum[0] = 1;
  
  // Loop over each container
  for (const container of containerSizes) {
    // Traverse backwards to avoid using same container twice
    for (let i = target; i >= container; i--) {
      waysToMakeSum[i] += waysToMakeSum[i - container];
    }
  }
  
  return waysToMakeSum[target];
}

function countMinPermutations(containerSizes: number[], target: number): {
  minContainers: number;
  permutationCount: number;
} {
  // First pass: find minimum containers needed
  const minContainersNeeded: number[] = new Array(target + 1).fill(Infinity);
  minContainersNeeded[0] = 0;
  
  for (const container of containerSizes) {
    for (let i = target; i >= container; i--) {
      minContainersNeeded[i] = Math.min(
        minContainersNeeded[i],
        minContainersNeeded[i - container] + 1
      );
    }
  }
  
  const minContainers = minContainersNeeded[target];
  
  // Second pass: count permutations using exactly minContainers
  const waysToMakeSumWithNContainers: number[][] = Array.from(
    { length: target + 1 },
    () => new Array(minContainers + 1).fill(0)
  );
  waysToMakeSumWithNContainers[0][0] = 1;
  
  for (const container of containerSizes) {
    for (let sum = target; sum >= container; sum--) {
      for (let numContainers = minContainers; numContainers >= 1; numContainers--) {
        waysToMakeSumWithNContainers[sum][numContainers] += 
          waysToMakeSumWithNContainers[sum - container][numContainers - 1];
      }
    }
  }
  
  return {
    minContainers,
    permutationCount: waysToMakeSumWithNContainers[target][minContainers]
  };
}

// ----------------------------------------------------------------------------------------------------
// | Part 1
// ----------------------------------------------------------------------------------------------------
function part1(input: string[]): number | undefined {
  let containerSizes: number[] = []
  let total = 150

  // Loop through our input and get the container sizes
  for (var i = 0; i < input.length; i++) {
    if (input.length == 0) {
      continue
    }

    let numToPush = parseInt(input[i])

    if (!isNaN(numToPush)) {
      containerSizes.push(numToPush)
    }
  }

  // Count the possible permutations
  let permutations: number = countPermutations(containerSizes, total);

  return permutations;
}

// ----------------------------------------------------------------------------------------------------
// | Part 2
// ----------------------------------------------------------------------------------------------------
function part2(input: string[]): number | undefined {
  let containerSizes: number[] = []
  let total = 150

  // Loop through our input and get the container sizes
  for (var i = 0; i < input.length; i++) {
    if (input.length == 0) {
      continue
    }

    let numToPush = parseInt(input[i])

    if (!isNaN(numToPush)) {
      containerSizes.push(numToPush)
    }
  }

  let permsForMinContainers = countMinPermutations(containerSizes, 150);

  return permsForMinContainers.permutationCount;
}


// ----------------------------------------------------------------------------------------------------
// | Solve the puzzle
// ----------------------------------------------------------------------------------------------------
function solve() {
  // Get the arguments
  const args = process.argv.slice(2);
  const fileName = args[0] || "input.txt";

  // Start the timer
  const startTimeStamp = new Date(Date.now()).toISOString();
  const startTime = performance.now();

  // Read the input as an array of strings
  const input = readInputAsArray(fileName);

// Run the parts
const p1Result = measurePerformance(() => part1(input));
const p2Result = measurePerformance(() => part2(input));

// End the timer
const endTimeStamp = new Date(Date.now()).toISOString();
const endTime = performance.now();
const duration = endTime - startTime;

// Return the results
return {
  Year: 2015,
  Day: 17,
  Part1: p1Result,
  Part2: p2Result,
  Duration: Math.round(duration * 1000000), // Convert to nanoseconds
  Timestamp: {
    Start: new Date(startTimeStamp).toISOString(),
    End: new Date(endTimeStamp).toISOString(),
  },
} as AOCDayResults;
  
}
const results = solve();
console.log(JSON.stringify(results, null, 2));